Changes for deploying smart contract on Publish Collection button

pracheebapate opened the pull request.

jibrail-idris reviewed (changes requested):
![image](https://user-images.githubusercontent.com/3234559/166654031-af53f877-1603-4951-aa01-668e3dccabc3.png)

This blockchain field marked in red should be resolved and hidden away from the UI. The value can be retrieved from the metamask provider chainId.

Also there should be two buttons as discussed. The first is the Draft button and the second one is publish.

jibrail-idris commented on src/pages/collections/single/Page.js:
```diff
@@ -46,6 +61,70 @@ const Page = () => {
+        // Connect to the network
+        let provider = ethers.getDefaultProvider('rinkeby');
+
+        // Load the wallet to deploy the contract with
```
The use of private key is wrong. This information should be encoded within the metamask provider.

pracheebapate replied:
Done. Removed private key and implemented signer.

pracheebapate commented:
> ![image](https://user-images.githubusercontent.com/3234559/166654031-af53f877-1603-4951-aa01-668e3dccabc3.png)
> 
> This blockchain field marked in red should be resolved and hidden away from the UI. The value can be retrieved from the metamask provider chainId.
> 
> Also there should be two buttons as discussed. The first is the Draft button and the second one is publish.
@jibrail-idris 
Noted. I will do that changes regarding blockchain filed.
Publish is currently placed on view single collection page currently.
![help2-2022-05-04_17-23](https://user-images.githubusercontent.com/103410278/166655226-917e1dab-f463-4d4f-89f1-dd287963243a.png)

pracheebapate commented:
> ![image](https://user-images.githubusercontent.com/3234559/166654031-af53f877-1603-4951-aa01-668e3dccabc3.png)
> 
> This blockchain field marked in red should be resolved and hidden away from the UI. The value can be retrieved from the metamask provider chainId.
> 
> Also there should be two buttons as discussed. The first is the Draft button and the second one is publish.

@jibrail-idris I have moved the 'Publish' button and deploy functionality to the 'Creat collection' page. Removed it from the view single collection page.

shakil-adv commented on src/components/cards/single/Card.js:
```diff
@@ -146,7 +146,7 @@ const Card = ({ data, type, variant }) => {
                                 logo: data.get("imageUrl"),
                                 collectionName: data.get("name"),
-                                itemCount: data.get("nftIds").length,
+                                itemCount: data.get("nftIds") ? data.get("nftIds").length : 0,
```
you can do `data.get('nftIds')?.length`

pracheebapate replied:
Done.

shakil-adv commented on src/pages/collections/single/Create.js:
```diff
@@ -305,20 +302,144 @@ const OwnerAddress = () => {
+
+            // Deploy ERC721 contract
+            // Connect to the network and get signer
+            const provider = new ethers.providers.Web3Provider(window.ethereum);
```
I think it's safe to check if `window.ethereum` object is present or not

shakil-adv commented on src/pages/collections/single/Create.js:
```diff
@@ -357,15 +478,16 @@ const CreateCollectionPage = () => {
 
         // save collection into the database
-        const {status, errorMessage} = await createCollection({
+        const {status, errorMessage, collectionId:newId} = await createCollection({
```
Please refer to https://apteam.atlassian.net/browse/DEVCVL-250 for the shape of `CvlCollection` table. `name`, `description`, `backgroundUrl` and `imageUrl` are stored in a IPFS (json file) and the link will be in the `metaDataURI` field

pracheebapate replied:
@shakil-adv 

CvlCollections: {
    id,
    name,
    blockchain,
    standard,
    network,
    categories (an array of pointers of category Objects)
    contractAddress,
    ownerAddress,
    metaDataURI,
    nftIds (an array of pointers of nft Objects)
}

I think we should store description on CvlCollections table as well. 

For metaDataURI  , json file (name, description, backgroundUrl and imageUrl) , I will update code .

pracheebapate replied:
@jibrail-idris @kyawthetkt @shakil-adv 
Had a discussion with Shakil today, and we have decided :
1. Keep description, backgroundUrl and imageUrl columns in the CvlCollections table
2. We don't require to create JSON file for collection metadata and store it in the field metaDataURI of CvlCollections table

Code changes are done for the same.

jibrail-idris replied:
@pracheebapate Yes, thanks. That is inline with our current understanding.

jibrail-idris commented on src/pages/collections/single/Create.js:
```diff
@@ -305,20 +302,144 @@ const OwnerAddress = () => {
+                    objectId:newCollectionId,
+                    contractAddress:contractAddress
+                });
+
```
@pracheebapate Can you show the contract address on the UI? The contract address can be a link back to the explorer.

jibrail-idris replied:
This is necessary because as a NFT minter I want to know the location of my deployed contract.

pracheebapate replied:
You can check this contract on rinkeby explorer.
**contract address: 0xeF61F7F18fD00b45250E7e4CAE04461ee3a00C64
contract hash: 0xd3f183c42a8cb94d212b9d67f827cd8ecfe41aa09787ab7017cf146483e78b32**

Videos and screenshots are attached.

https://user-images.githubusercontent.com/103410278/166882578-84dc872d-fe12-4ef3-b286-941dc668754e.mp4

![for-jib-3-2022-05-05_15-51](https://user-images.githubusercontent.com/103410278/166882590-9eacdd18-428f-46bf-8bbb-070cf4c0e988.png)
![for-jib-2-2022-05-05_15-51](https://user-images.githubusercontent.com/103410278/166882593-519d59e0-c452-407e-aea5-cf718008d316.png)
![for-jib-1-2022-05-05_15-51](https://user-images.githubusercontent.com/103410278/166882595-a5440439-e0f8-4ee1-90db-33eaffd3d7b4.png)

pracheebapate replied:
Let me check where and how to show the contract address on the create collection page. We can also show the address and explorer link on 'View Collection' page.

Showing the explorer link will depend on the network used by the user so not sure about that.

pracheebapate replied:
@jibrail-idris Done. 
Once the contract is deployed and collection is published, the contract address and explorer link will be shown on the same page.
![for-jib-comment=2022-05-05_18-57](https://user-images.githubusercontent.com/103410278/166910184-69d790ed-51e6-4284-9b30-2962641a0018.png)

kyawthetkt commented on src/pages/collections/single/Create.js:
```diff
@@ -1,11 +1,14 @@
 import useCvlCollectionData from '../../../databases/moralis/useCvlCollectionData';
+import { ethers } from "ethers";
+import CVL_ERC721 from "../../../contracts/abi/CVL_ERC721.json"
+import Processing from "../../../components/processing/Processing";
```
`import CVL_ERC721 from "../../../contracts/abi/CVL_ERC721.json`

Instead of using complex relative paths,you could use
`import CVL_ERC721 from "contracts/abi/CVL_ERC721.json"`

as inside jsconfig.json of root dir,we define like so
`{
    "compilerOptions": {
      "baseUrl": "src"
    },
    "include": ["src"]
  }`.

pracheebapate replied:
Done. Thanks for this suggestion.
